components/sync_manager/ios/SyncManager/SyncManagerTelemetry.swift

```diff
@@ -0,0 +1,358 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Glean
+import UIKit
+
+enum SupportedEngines: String {
+    case History = "history"
+    case Bookmarks = "bookmarks"
+    case Logins = "passwords"
+    case CreditCards = "creditcards"
+    case Tabs = "tabs"
+}
+
+enum TelemetryReportingError: Error {
+    case InvalidEngine(message: String)
+    case UnsupportedEngine(message: String)
+}
+
+func processSyncTelemetry(syncTelemetry: RustSyncTelemetryPing,
+                          submitGlobalPing: (NoReasonCodes?) -> Void = GleanMetrics.Pings.shared.sync.submit,
+                          submitHistoryPing: (NoReasonCodes?) -> Void = GleanMetrics.Pings.shared.historySync.submit,
+                          submitBookmarksPing: (NoReasonCodes?) -> Void = GleanMetrics.Pings.shared.bookmarksSync.submit,
+                          submitLoginsPing: (NoReasonCodes?) -> Void = GleanMetrics.Pings.shared.loginsSync.submit,
+                          submitCreditCardsPing: (NoReasonCodes?) -> Void = GleanMetrics.Pings.shared.creditcardsSync.submit,
+                          submitTabsPing: (NoReasonCodes?) -> Void = GleanMetrics.Pings.shared.tabsSync.submit) throws
+{
+    for syncInfo in syncTelemetry.syncs {
+        _ = GleanMetrics.Sync.syncUuid.generateAndSet()
+
+        if let failureReason = syncInfo.failureReason {
+            recordFailureReason(reason: failureReason,
+                                failureReasonMetric: GleanMetrics.Sync.failureReason)
+        }
+
+        try syncInfo.engines.forEach { engineInfo in
+            switch engineInfo.name {
+            case SupportedEngines.Bookmarks.rawValue:
+                try individualBookmarksSync(hashedFxaUid: syncTelemetry.uid,
```

Comment:
nit: How do we feel about introducing a generic "reportEngineSync" function, that's generic over some protocol that defines a sync engines telemetry? The protocol can define what it means for an engine to report its metrics, then we can have a type per engine implementing it


Thoughts? This isn't bad either, so I don't feel strongly!

Reply:
That would definitely be an improvement. Making this change now really depends on how long this code will be in use. I saw it as a stepping stone towards having all our sync telemetry logic in rust. But if that's a ways off, this might be a worthwhile change. 

I'll talk to Sammy and see what his telemetry plans are since I think he owns the epic. If it makes sense, I'll timebox making this change tomorrow.

Reply:
Whoops, meant to circle back to this yesterday. For posterity, I'm going to leave this logic as is because 1) Sammy may begin the work to replace this code as early as May and 2) I don't think this code suffers to much without the changes.